sdk/core/azure-core/src/cryptography/md5.cpp

```diff
@@ -16,156 +16,195 @@
 #include <stdexcept>
 #include <vector>
 
-namespace Azure { namespace Core { namespace Cryptography {
+namespace {
 
 #if defined(AZ_PLATFORM_WINDOWS)
 
-  namespace _detail {
-    struct AlgorithmProviderInstance
+void ThrowIfFails(NTSTATUS status)
+{
+  if (!NT_SUCCESS(status))
+  {
+    throw std::runtime_error("BCryptOpenAlgorithmProvider failed");
+  }
+}
+
+class Md5AlgorithmProvider {
+private:
+  NT_STATUS m_status = STATUS_UNSUCCESSFUL;
+
+public:
+  BCRYPT_ALG_HANDLE Handle;
+  std::size_t ContextSize;
+  std::size_t HashLength;
+
+  Md5AlgorithmProvider()
+  {
+    // open an algorithm handle
+    if (!NT_SUCCESS(
+            m_status = BCryptOpenAlgorithmProvider(&Handle, BCRYPT_MD5_ALGORITHM, nullptr, 0)))
```

Comment:
Can we use `ThrowIfFails` here and elsewhere?

Reply:
Not if we want to display a message next to the code.  Because the message would need to be created even before we throw